docs/proposals/predictive-horizontal-pod-autoscaling.md

```diff
@@ -0,0 +1,232 @@
+initialize, the already existing pods will spend the entire
+initialization time operating outside of target
+resource consumption while they wait for the newly created pods to
+share the computational work. Depending on pod initialization time, and the
+danger of operating outside target resource consumption, it can be
+extremely detrimental to not account for pod initialization time.
+
+**Predictive horizontal pod autoscaling** addresses this problem.
+It accounts for pod initialization time by
+determining the target number of pods based on the predicted future pod resource
+consumption, instead of the current pod resource consumption. In other words,
+if it takes 10 minutes for a pod to initialize, than if it is 6:00pm and
+predictive autoscaling is enabled, autoscaling behavior is determined by a
+prediction of the application’s demands at 6:10pm. If it is determined that a
+new instance of this pod is needed based on the prediction, the new pod will be
+created at 6:00pm, meaning it will be initialized by 6:10pm. The pods will
+never have to operate outside of their target resource consumption (as opposed
+to if autoscaling was occurring in a non-predictive manner, in which case the
+pod would not be created until 6:10 and would not be initialized until 6:20pm,
+meaning there was 10 minutes in which pods were operating outside of target
+resource consumption). Predictive horizontal pod autoscaling can
+decrease the impact of pod initialization times on autoscaling responsiveness and thus
+decrease the amount of time pods spend operating outside of their target
+resource consumption.
+
+## Predictive Autoscaling Algorithm
+
+Currently, horizontal pod autoscaling is implemented as a control loop.
+Every period, the autoscaler checks all machines specified by the
+`Status.PodSelector` of the Scale subresource, and checks their metrics through
+the [Compute Resource Metrics Api](compute-resource-metrics-api.md).
+It computes the `TargetNumOfPods` based on the previously described algorithm, with the following:
+
+```
+TargetNumOfPods = ceil(sum(CurrentPodsResourceConsumption) / TargetResourceConsumption)
+```
+
+Predictive horizontal autoscaling will use a similar algorithm with one
+significant difference. Rather than calculating based on the
+`CurrentPodsResourceConsumption`, we will calculate based on
+`FuturePodsResourceConsumption`, which we initially define as
+`FuturePodsResourceConsumption = CurrentPodResourceConsumption +
+(PodsAverageInitializationTime * DerivativeCurrentPodsResourceConsumptionPerSecond)`.
+The calculation for `FuturePodsResourceConsumption` could become more complex,
+incorporating increasingly complex methods of prediction, in the future.
+We now calculate `TargetNumOfPods` with the following:
+
+```
+TargetNumOfPods = ceil(sum(FuturePodsResourceConsumption)) / Target
+```
+
+The modifications for this option will predominantly occur in
+`pkg/controllers/podautoscaler/horizontal.go` in the
```

Comment:
My original plan was to use an annotation on the auto-scaler object so that prediction could easily be turned on and off. And within the `pkg/controllers/podautoscaler/horizontal.go` file I would define a `predictiveAutoscaling` function that would be called if the hpa object was annotated to turn on prediction. Does this fit within the idea of having an addon as opposed to a modification?

I guess my only worry about creating a separate predictive hpa controller is that there would be a lot of code duplication between the two. But I also maybe missing a better way of doing this.